### PR TITLE
For #4399: Create LibState Controller for Search

### PR DESCRIPTION
**Pull request by dragneelfps**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/fenix/blob/master/CHANGELOG.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

**Comment by sblatz** [comment]
@dragneelfps this is looking really awesome so far. Thanks so much! 

Are you interested in writing tests for the controller as well? If so, that should be part of this PR as we're trying to land tests with these controller updates to make sure no functionality is breaking :)

If not I'm happy to add on to this with some tests!

**Comment by dragneelfps** [comment]
Sure. I'll add some tests for the controller.

**Comment by dragneelfps** [comment]
@sblatz Is there any guidelines for testing private methods? For instance, how should I go about testing method `createSearchEvent()` of `DefaultSearchController`. Or if we don't test private methods at all?

**Comment by sblatz** [comment]
@dragneelfps if you think a private method needs tested, feel free to change it to an `internal` function 😄

**Comment by boek** [comment]
We might also be able to try: https://developer.android.com/reference/kotlin/androidx/annotation/VisibleForTesting?authuser=1

:)

**Comment by sblatz** [comment]
@dragneelfps I'm going to review this today and hopefully land it. Even if we don't have *all* the tests I really don't want this PR to rot and have merge conflicts.

**Comment by dragneelfps** [comment]
@sblatz I have rebased my pr onto the master and now the task cluster is passing.

File: app/src/test/java/org/mozilla/fenix/search/DefaultSearchControllerTest.kt [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            engine = searchEngine,
+            forceSearch = true
+        ) }
+    }
+
+    @Test
+    fun handleSearchShortcutEngineSelected() {
+        val searchEngine: SearchEngine = mockk(relaxed = true)
+
+        println("HELLo: ${searchEngine.name}")
```
Comment (sblatz) [inline app/src/test/java/org/mozilla/fenix/search/DefaultSearchControllerTest.kt]: Let's get rid of this :)

File: app/src/test/java/org/mozilla/fenix/search/DefaultSearchControllerTest.kt [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import org.mozilla.fenix.HomeActivity
+import org.mozilla.fenix.R
+import org.mozilla.fenix.components.metrics.Event
+import org.mozilla.fenix.components.metrics.MetricController
+import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.ext.metrics
+import org.mozilla.fenix.ext.nav
+import org.mozilla.fenix.ext.searchEngineManager
+
+class DefaultSearchControllerTest {
```
Comment (sblatz) [inline app/src/test/java/org/mozilla/fenix/search/DefaultSearchControllerTest.kt]: These tests all look awesome, thank you for adding!! 🎉

File: app/src/test/java/org/mozilla/fenix/search/DefaultSearchControllerTest.kt (hunk 1 above)
Comment (dragneelfps) [reply app/src/test/java/org/mozilla/fenix/search/DefaultSearchControllerTest.kt]: done

**Comment by dragneelfps** [comment]
@sblatz The regression is gone after rebasing onto master once again. I think the regression was introduced from somewhere else.

**Comment by sblatz** [comment]
@dragneelfps awesome! I manually tested and am not seeing this issue anymore either 😄

**Pull request merged**